### PR TITLE
V3 agent changes 

### PR DESCRIPTION
**Pull request by daniel-savu**
### Description

Closes https://github.com/hyperlane-xyz/issues/issues/561

Builds on top of https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2740

Depends on https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2681 for e2e testing

This PR includes:
- [x] Merkle tree hook indexer
- [x] Merkle tree builder task
- [x] Update submitter to trigger retries if no proof is available yet

Slightly more detailed overview of the work here: https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2720#issuecomment-1724038643

### CODE REVIEW

File: rust/hyperlane-base/src/db/rocks/hyperlane_db.rs [hunk 1]
```diff
@@ -152,6 +154,23 @@ impl HyperlaneRocksDB {
         Ok(true)
     }
 
+    /// Store the merkle tree insertion event, and also store a mapping from message_id to leaf_index
+    pub fn process_tree_insertion(&self, insertion: &MerkleTreeInsertion) -> DbResult<bool> {
+        if let Ok(Some(_)) = self.retrieve_merkle_leaf_index_by_message_id(&insertion.message_id())
```
Comment (aroralanuk) [inline rust/hyperlane-base/src/db/rocks/hyperlane_db.rs:159]: isn't this not allowing double insertions?

File: rust/hyperlane-base/src/db/rocks/hyperlane_db.rs [hunk 2]
```diff
@@ -152,6 +154,23 @@ impl HyperlaneRocksDB {
...
+        {
+            debug!(insertion=?insertion, "Tree insertion already stored in db");
+            return Ok(false);
+        }
+        // even if double insertions are ok, store the leaf by `leaf_index` (guaranteed to be unique)
+        // rather than by `message_id` (not guaranteed to be unique), so that leaves can be retrieved
+        // based on insertion order.
+        self.store_merkle_tree_insertion_by_leaf_index(&insertion.index(), insertion)?;
```
Comment (aroralanuk) [inline rust/hyperlane-base/src/db/rocks/hyperlane_db.rs:167]: where are we checking for uniqueness per leaf_index?

shouldn't `retrieve_merkle_message_id_by_leaf_index(leaf_index) == 0 || retrieve_merkle_message_id_by_leaf_index(leaf_index) == insertion.message_id()` be necessary too here?

File: rust/hyperlane-base/src/db/rocks/hyperlane_db.rs (hunk 1 above)
Comment (daniel-savu) [reply rust/hyperlane-base/src/db/rocks/hyperlane_db.rs:159]: great point, yes

File: rust/agents/relayer/src/merkle_tree/processor.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            // behind a shared Arc.
+        } else {
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+}
+
+impl MerkleTreeProcessor {
+    fn next_unprocessed_leaf(&mut self) -> Result<Option<MerkleTreeInsertion>> {
```
[the remark this answers is not in the document]
Comment (daniel-savu) [reply rust/agents/relayer/src/merkle_tree/processor.rs:68]: it means not inserted into the prover_sync, open to better naming suggestions

**Comment by daniel-savu** [comment]
Closing in favour of https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2746

**Pull request closed without merging**